tokio-trace-core/src/field.rs

```diff
@@ -56,8 +56,24 @@
 //! indefinitely, they are not heap-allocated by default, to avoid unnecessary
 //! allocations, but the `IntoValue` trait presents `Subscriber`s with the
 //! _option_ to box values should they need to do so.
+use super::Meta;
 use std::{any::TypeId, borrow::Borrow, fmt};
 
+/// Trait implemented to allow a type to be used as a field key.
+///
+/// **Note**: Although this is implemented for both the [`Key`] type *and* any
+/// type that can be borrowed as an `&str`, only `Key` allows _O_(1) access.
+/// Indexing a field with a string results in an iterative search that performs
+/// string comparisons. Thus, if possible, once the key for a field is known, it
+/// should be used whenever possible.
+pub trait AsKey {
```

Comment:
Why is this trait necessary?

Reply:
The motivation behind this is that it let's functions that index a field take a `Key` _or_ a `&str`, which seemed like a more ergonomic API than having two separate methods (I.e. `add_field_by_name`/`add_field_by_key`). The std conversion traits won't work in that case, as converting a string to a `Key` requires access to the metadata.

Reply:
IMO `-core` should favor less traits / types / code over ergonomics. We can revisit ergonomics in the other crates.

Reply:
Okay, I'll make the field indexing in `core` by key only, and move this to `tokio-trace`.